Microsoft.Azure.Cosmos/src/HttpClient/CosmosHttpClient.cs

```
@@ -14,16 +14,42 @@ internal abstract class CosmosHttpClient : IDisposable
     {
         public abstract HttpMessageHandler HttpMessageHandler { get; }
 
+        public enum TimeoutPolicy
+        {
+            Standard,
+            ControlPlane,
+            ControlPlaneHotPath,
+        }
+
+        public static TimeoutPolicy GetTimeoutPolicy(
+            DocumentServiceRequest documentServiceRequest)
+        {
+            if (documentServiceRequest.ResourceType == ResourceType.Document
+                && documentServiceRequest.OperationType == OperationType.QueryPlan)
+            {
+                return TimeoutPolicy.ControlPlaneHotPath;
```

This is going to increase the number of timeouts for query plan. Without extensive diagnostics query is going to get all the CRIs.

---

1. The CosmosDiagnostics already includes the extensive diagnostics which records each failure and retry. This is already wired through with query diagnostics.
2. This is going to decrease the number of timeouts for query plan. It now fails fast, and has retry logic for the timeout failures. Before it would wait 65 seconds before timing out. Now it will retry .5 sec, 5 sec, 10 seconds to get the query plan which is way faster than the previous waiting of 65 seconds.

---

If a query plan request takes +10 seconds, isn't that an issue that needs to be investigated anyway?

---

Can you send an example diagnostics when we hit a timeout from the gateway getting query plan? From there we can determine if it's dead obvious that the gateway failed. Otherwise query is just going to get a bunch of CRIs for timeouts it doesn't own.